library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,190 @@
+    let mut result = [(0, 0); 256];
+
+    let mut sqrt = 0;
+    let mut i = 0;
+    'outer: loop {
+        let mut remaining = 2 * sqrt + 1;
+        while remaining > 0 {
+            result[i as usize] = (sqrt, 2 * sqrt + 1 - remaining);
+            i += 1;
+            if i >= result.len() {
+                break 'outer;
+            }
+            remaining -= 1;
+        }
+        sqrt += 1;
+    }
+
+    result
+};
+
+// `#[inline(always)]` because the programmer-accessible functions will use
+// this internally and the contents of this should be inlined there.
+#[inline(always)]
+pub const fn u8(n: u8) -> u8 {
+    ISQRT_AND_REMAINDER_8_BIT[n as usize].0
+}
+
+#[inline(always)]
+const fn intermediate_u8(n: u8) -> (u8, u8) {
+    ISQRT_AND_REMAINDER_8_BIT[n as usize]
+}
+
+macro_rules! karatsuba_isqrt {
+    ($FullBitsT:ty, $fn:ident, $intermediate_fn:ident, $HalfBitsT:ty, $half_fn:ident, $intermediate_half_fn:ident) => {
+        // `#[inline(always)]` because the programmer-accessible functions will
+        // use this internally and the contents of this should be inlined
+        // there.
+        #[inline(always)]
+        pub const fn $fn(mut n: $FullBitsT) -> $FullBitsT {
+            // Performs a Karatsuba square root.
+            // https://web.archive.org/web/20230511212802/https://inria.hal.science/inria-00072854v1/file/RR-3805.pdf
+
+            const HALF_BITS: u32 = <$FullBitsT>::BITS >> 1;
+            const QUARTER_BITS: u32 = <$FullBitsT>::BITS >> 2;
+
+            let leading_zeros = n.leading_zeros();
+            let result = if leading_zeros >= HALF_BITS {
+                $half_fn(n as $HalfBitsT) as $FullBitsT
+            } else {
+                // Either the most-significant bit or its neighbor must be a one, so we shift left to make that happen.
+                let precondition_shift = leading_zeros & (HALF_BITS - 2);
+                n <<= precondition_shift;
+
+                let hi = (n >> HALF_BITS) as $HalfBitsT;
+                let lo = n & (<$HalfBitsT>::MAX as $FullBitsT);
+
+                let (s_prime, r_prime) = $intermediate_half_fn(hi);
+
+                let numerator = ((r_prime as $FullBitsT) << QUARTER_BITS) | (lo >> QUARTER_BITS);
+                let denominator = (s_prime as $FullBitsT) << 1;
+
+                let q = numerator / denominator;
+                let u = numerator % denominator;
+
+                let mut s = (s_prime << QUARTER_BITS) as $FullBitsT + q;
+                if ((u << QUARTER_BITS) | (lo & ((1 << QUARTER_BITS) - 1))) < q * q {
+                    s -= 1;
+                }
+                s >> (precondition_shift >> 1)
+            };
+
+            result
+        }
+
+        const fn $intermediate_fn(mut n: $FullBitsT) -> ($FullBitsT, $FullBitsT) {
+            // Performs a Karatsuba square root.
+            // https://web.archive.org/web/20230511212802/https://inria.hal.science/inria-00072854v1/file/RR-3805.pdf
```

Comment:
```suggestion
        /// Perform a Karatsuba square root without discarding the remainder.
        ///
        /// https://web.archive.org/web/20230511212802/https://inria.hal.science/inria-00072854v1/file/RR-3805.pdf
        const fn $intermediate_fn(mut n: $FullBitsT) -> ($FullBitsT, $FullBitsT) {
```

Or some kind of note to say how `$intermediate_fn` is different from `$fn`.

Reply:
I've added rustdocs to everything in `library/core/src/num/int_sqrt.rs`, but I've also eliminated the `intermediate` functions by manually inlining them.